nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs

```diff
@@ -0,0 +1,843 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A mechanism for allocating clickhouse keeper and server nodes for clustered
+//! clickhouse setups during blueprint planning
+
+use clickhouse_admin_types::KeeperId;
+use nexus_types::deployment::{
+    all_omicron_zones, BlueprintZoneFilter, BlueprintZoneType,
+    BlueprintZonesConfig, ClickhouseClusterConfig,
+};
+use nexus_types::inventory::ClickhouseKeeperClusterMembership;
+use omicron_uuid_kinds::{OmicronZoneUuid, SledUuid};
+use slog::{error, Logger};
+use std::collections::{BTreeMap, BTreeSet};
+use thiserror::Error;
+
+// The set of in service clickhouse server and keeper zones as constructed by
+// the `BlueprintBuilder` in the current planning iteration.
+//
+// Will be removed once the planner starts using this code
+#[allow(unused)]
+struct InServiceClickhouseZones {
+    keepers: BTreeSet<OmicronZoneUuid>,
+    servers: BTreeSet<OmicronZoneUuid>,
+}
+
+impl From<&BTreeMap<SledUuid, BlueprintZonesConfig>>
+    for InServiceClickhouseZones
+{
+    fn from(
+        zones_by_sled_id: &BTreeMap<SledUuid, BlueprintZonesConfig>,
+    ) -> Self {
+        let mut keepers = BTreeSet::new();
+        let mut servers = BTreeSet::new();
+        for (_, bp_zone_config) in all_omicron_zones(
+            zones_by_sled_id,
+            BlueprintZoneFilter::ShouldBeRunning,
+        ) {
+            match bp_zone_config.zone_type {
+                BlueprintZoneType::ClickhouseKeeper(_) => {
+                    keepers.insert(bp_zone_config.id);
+                }
+                BlueprintZoneType::ClickhouseServer(_) => {
+                    servers.insert(bp_zone_config.id);
+                }
+                _ => (),
+            }
+        }
+        InServiceClickhouseZones { keepers, servers }
+    }
+}
+
+/// A mechanism for allocating clickhouse server and keeper processes
+/// to zones
+///
+/// This is to be used as part of the `BlueprintBuilder` after zones have been
+/// allocated.
+//
+// Will be removed once the planner starts using this code
+#[allow(unused)]
+pub struct ClickhouseAllocator {
+    log: Logger,
+    in_service_clickhouse_zones: InServiceClickhouseZones,
+    parent_config: ClickhouseClusterConfig,
+    // The latest clickhouse cluster membership from inventory
+    inventory: Option<ClickhouseKeeperClusterMembership>,
```

Comment:
When would ClickhouseKeeperClusterMembership be `None`? (This question is just me trying to understand how reconfigurator works.)

Reply:
It would be `None` if the latest inventory collection failed to reach all keepers. This would most likely happen due to a transient error like a network partition or reboot.

Reply:
And since reconfigurator is constantly polling (is this right?), it's not a big deal if this information is empty every so often?